### ISSUE TITLE
Decorator doesn't work on adapter

### ISSUE THREAD

**Issue by YairHalberstadt** [autofac/Autofac#1004]
Autofac version: 4.9.2

Here is a failing unit test reproducing the issue:

```csharp
using Xunit;

public class Tests
{
		public class Foo { }
		public class Bar : IBar { }
		public interface IBar { }
		public class BarDecorator : IBar { public BarDecorator(IBar bar) { } }

		[Fact]
		public void DecoratorWorksOnAdapter()
		{
			var builder = new ContainerBuilder();
			builder.RegisterType<Foo>();
			builder.RegisterAdapter<Foo, IBar>(f => new Bar());
			builder.RegisterDecorator<BarDecorator, IBar>();
			var container = builder.Build();
			Assert.IsType<BarDecorator>(container.Resolve<IBar>());
		}
}
```

    Assert.IsType() Failure
    Expected: Tests+BarDecorator
    Actual: Tests+Bar

**Comment by tillig** [comment]
This is an interesting issue. Have you tried [the classic decorator syntax](https://autofac.readthedocs.io/en/latest/advanced/adapters-decorators.html#classic-syntax) to see what happens? I'd be curious if it works the old way and not the new way.

I started trying to think if there was a way someone could get themselves stuck, like if they somehow wired up a circular chain of component => adapter => decorator => adapter where the two adapters there are somehow the same, but I couldn't actually come up with anything, so combining the functionality should work.

I think in general the order of operations should be decorator-then-adapter (if that hasn't already been determined in the code).

```c#
builder.RegisterType<Foo>().As<IFoo>();
builder.RegisterDecorator<FooDecorator, IFoo>();
builder.RegisterAdapter<IFoo, IBar>(f => new Bar(f));
builder.RegisterDecorator<BarDecorator, IBar>();
builder.RegisterAdapter<IBar, IBaz>(b => new Baz(b));
```

Resolving `IBaz` should yield `Baz` => `BarDecorator` => `Bar` => `FooDecorator` => `Foo`.

**Comment by YairHalberstadt** [comment]
The following does indeed work:

```csharp
using Autofac;
using Xunit;

public class Tests
{
	public class Foo : IFoo { }
	public interface IFoo { }
	public class Bar : IBar { }
	public interface IBar { }
	public class BarDecorator : IBar { public BarDecorator(IBar bar) { } }

	[Fact]
	public void DecoratorWorksOnAdapter()
	{
		var builder = new ContainerBuilder();
		builder.RegisterType<Foo>().Named<IFoo>("foo");
		builder.RegisterAdapter<IFoo, IBar>(f => new Bar());
		builder.RegisterDecorator<IBar>((c, inner) => new BarDecorator(inner), fromKey: "foo");
		var container = builder.Build();
		Assert.IsType<BarDecorator>(container.Resolve<IBar>());
	}
}
```

**Comment by tillig** [comment]
Ah! Good to know. That narrows it down very specifically to the new decorator syntax, so it should be easier to start fixing it. Thanks for checking!

**Comment by YairHalberstadt** [comment]
InstanceDecorator.TryDecorateRegistration line 44 appears to be the culprit here:

```csharp
            // Issue #965: Do not apply the decorator if already applied, or if the registration is for an adapter.
            if (registration.IsAdapting() || registration.Activator.LimitType != instanceType) return instance;
```

When that line is removed, this use case works, but the following tests fail:

```
Autofac.Specification.Test.Features.DecoratorTests.CanResolveDecoratorWithFunc
Autofac.Specification.Test.Features.DecoratorTests.CanResolveDecoratorWithLazy
Autofac.Specification.Test.Features.DecoratorTests.DecoratorAppliedOnlyOnceToComponentWithExternalRegistrySource
Autofac.Specification.Test.Features.DecoratorTests.DecoratorCanBeAppliedTwiceInChildLifetimeScope
Autofac.Specification.Test.Features.DecoratorTests.DecoratorCanBeAppliedTwiceIntentionallyWithExternalRegistrySource
Autofac.Specification.Test.Features.DecoratorTests.DecoratorCanBeRegisteredInChildLifetimeScope
```

**Comment by alexmg** [comment]
This should be resolved with #1023.

**Comment by YairHalberstadt** [comment]
Thanks alexmg! Much appreciated!
